### PR TITLE
Fixes #15070. Allow custom algorithm ID ASN.1 encoding for provided ciphers

### LINKED ISSUE #15070: Some old algorithms in PKCS#12 files don't work any more
Using alpha16-dev (1f3b58d8413cfa) compiled with `./config enable-md2 enable-rc5 no-shared enable-ec_nistp_64_gcc_128 -ggdb` on Fedora 32.

Some files that are readable with 1.1.1 aren't readable with 3.0.0-alpha16-dev, even with legacy provider loaded.

Algorithms affected:
 - PBES2 with RC2-CBC-128
 - PBES2 with RC2-CBC-40
 - PBES2 with RC2-CBC-64
 - PBES1 with pbeWithSHAAnd128BitRC4
 - PBES1 with pbeWithSHAAnd40BitRC4

Reproducer for RC2-CBC-128:
```
# compile openssl with old algorithms enabled
# checkout the keyfile corpus in parallel directory:
git clone https://github.com/redhat-qe-security/keyfile-corpus.git
cd openssl
apps/openssl pkcs12 -provider-path providers/ -provider legacy -provider default -in ../'keyfile-corpus/rsa(2048,sha256),cert&key(PBES2(PBKDF2(salt(8),iter(2048),keyLen(16),prf(default)),rc2-cbc(keyBits(56=128bit),IV(8)))),mac(sha1,salt(8),iter(2048)),pass(ascii).p12' -out /dev/null -info -nodes -passin file:../keyfile-corpus/password-ascii.txt
```

OpenSSL output:
```
MAC: sha1, Iteration 2048
MAC length: 20, salt length: 8
PKCS7 Encrypted data: PBES2, PBKDF2, RC2-CBC, Iteration 2048, PRF hmacWithSHA1
Error outputting keys and certificates
4057DE1C127F0000:error:0680006D:asn1 encoding routines:ASN1_TYPE_get_octetstring:data is wrong:crypto/asn1/evp_asn1.c:37:
4057DE1C127F0000:error:0300007A:digital envelope routines:evp_cipher_asn1_to_param_ex:cipher parameter error:crypto/evp/evp_lib.c:230:
4057DE1C127F0000:error:0300007A:digital envelope routines:PKCS5_v2_PBE_keyivgen:cipher parameter error:crypto/evp/p5_crpt2.c:149:
```

Reproducer for RC4:
```
apps/openssl pkcs12 -provider-path providers/ -provider legacy -provider default -in '../keyfile-corpus/rsa(2048,sha256),cert&key(pbeWithSHAAnd128BitRC4,salt(8),iter(2048)),mac(sha1,salt(8),iter(2048)),pass(ascii).p12' -out /dev/null -info -nodes -passin 'file:../keyfile-corpus/password-ascii.txt'
```
OpenSSL output:
```
MAC: sha1, Iteration 2048
MAC length: 20, salt length: 8
PKCS7 Encrypted data: pbeWithSHA1And128BitRC4, Iteration 2048
Error outputting keys and certificates
40374D13F17F0000:error:1180006A:PKCS12 routines:PKCS12_PBE_keyivgen:iv gen error:crypto/pkcs12/p12_crpt.c:61:
```

Other reproducers:
```
apps/openssl pkcs12 -provider-path providers/ -provider legacy -provider default -in '../keyfile-corpus/rsa(2048,sha256),cert&key(PBES2(PBKDF2(salt(8),iter(2048),keyLen(5),prf(default)),rc2-cbc(keyBits(160=40bit),IV(8)))),mac(sha1,salt(8),iter(2048)),pass(ascii).p12' -out /dev/null -info -nodes -passin 'file:../keyfile-corpus/password-ascii.txt'
apps/openssl pkcs12 -provider-path providers/ -provider legacy -provider default -in '../keyfile-corpus/rsa(2048,sha256),cert&key(PBES2(PBKDF2(salt(8),iter(2048),keyLen(8),prf(default)),rc2-cbc(keyBits(120=64bit),IV(8)))),mac(sha1,salt(8),iter(2048)),pass(ascii).p12' -out /dev/null -info -nodes -passin 'file:../keyfile-corpus/password-ascii.txt'
apps/openssl pkcs12 -provider-path providers/ -provider legacy -provider default -in '../keyfile-corpus/rsa(2048,sha256),cert&key(pbeWithSHAAnd40BitRC4,salt(8),iter(2048)),mac(sha1,salt(8),iter(2048)),pass(ascii).p12'  -out /dev/null -info -nodes -passin 'file:../keyfile-corpus/password-ascii.txt'
```

### PR DESCRIPTION
**Pull request by jon-oracle**
Allows custom ASN.1 encode/decode of algorithm parameters for provided ciphers, only used by RC2 currently.

Also adds a few extra ciphers to the PKCS#12 PBES2 format test.

- [X] tests are added or updated

### CODE REVIEW

File: crypto/evp/evp_lib.c [hunk 1]
```diff
@@ -369,6 +369,11 @@ int evp_cipher_cache_constants(EVP_CIPHER *cipher)
         /* Provided implementations may have a custom cipher_cipher */
         if (cipher->prov != NULL && cipher->ccipher != NULL)
             cipher->flags |= EVP_CIPH_FLAG_CUSTOM_CIPHER;
+        /* Provided implementations may also have custom ASN1 algorithm parameters */
+        if (EVP_CIPHER_gettable_ctx_params(cipher) != NULL
```
Comment (paulidale) [inline crypto/evp/evp_lib.c:373]: The condition only needs to be:

`OSSL_PARAM_locate_const(EVP_CIPHER_gettable_ctx_params(cipher), OSSL_CIPHER_PARAM_ALG_ID) != NULL`

The locate functions handle NULL properly.

File: providers/implementations/ciphers/cipher_rc2.c [hunk 2]
```diff
@@ -210,11 +210,13 @@ static int rc2_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 
 CIPHER_DEFAULT_GETTABLE_CTX_PARAMS_START(rc2)
 OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_RC2_KEYBITS, NULL),
+OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_ALG_ID, NULL, 0),
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_rc2.c:213]: Yes- this was definitely a bug, considering that this is the only cipher that uses this ID.

File: test/pkcs12_format_test.c [hunk 3]
```diff
@@ -733,6 +791,10 @@ int setup_tests(void)
     ADD_TEST(test_cert_key_with_attrs_and_mac);
     ADD_TEST(test_cert_key_encrypted_content);
     ADD_TEST(test_single_secret_encrypted_content);
+    if (lgcyprov == NULL)
```
Comment (slontis) [inline test/pkcs12_format_test.c:794]: I would probably move this up to group it with the other test case that is like this.

**Review by slontis — COMMENTED** [review]
Good work figuring this out....

**Comment by levitte** [comment]
I've pondered if we shouldn't pass the full AlgorithmIdentifier across, just for consistency with everything else that does this (grep for ALGORITHM_ID).
Another thing I've pondered is that ALG_ID is not a good macro name for just the AlgorithmIdentifier params field.  Considering ALGORITHM_ID, what about ALGORITHM_ID_PARAMS?

**Comment by jon-oracle** [comment]
> I've pondered if we shouldn't pass the full AlgorithmIdentifier across, just for consistency with everything else that does this (grep for ALGORITHM_ID).

Fair point, but that's a bigger chunk of work. The EVP_CIPHER_param_to_asn1()/EVP_CIPHER_asn1_to_param() APIs will still require handling just the parameters.

> Another thing I've pondered is that ALG_ID is not a good macro name for just the AlgorithmIdentifier params field. Considering ALGORITHM_ID, what about ALGORITHM_ID_PARAMS?

OSSL_CIPHER_PARAM_ALGORITHM_ID_PARAMS is a little long but is a more accurate/readable name.

File: include/openssl/core_names.h [hunk 4]
```diff
@@ -94,7 +94,7 @@ extern "C" {
 #define OSSL_CIPHER_PARAM_SPEED                "speed"        /* uint */
 #define OSSL_CIPHER_PARAM_CTS_MODE             "cts_mode"     /* utf8_string */
 /* For passing the AlgorithmIdentifier parameter in DER form */
-#define OSSL_CIPHER_PARAM_ALG_ID               "alg_id_param" /* octet_string */
+#define OSSL_CIPHER_PARAM_ALGORITHM_ID_PARAMS  "alg_id_param" /* octet_string */
```
Comment (slontis) [inline include/openssl/core_names.h:97]: update doc/man7/provider-cipher.pod entry for this also. (search for alg_id_param)

Comment (jon-oracle) [reply include/openssl/core_names.h:97]: Done

**Comment by slontis** [comment]
Apart from the checksum this LGTM..

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
Rebase + make update needed

**Review by slontis — APPROVED** [review]
Approved assuming it is rebased + make updated

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Fixed the checksums during merge.

Merged to master. Thank you for the contribution!

**Pull request closed without merging**